src/probability/martingale/borel_cantelli.lean

```diff
@@ -0,0 +1,400 @@
+      simp only [neg_neg, pi.neg_apply] } },
+  rw [hω₁, this, ← hω₂],
+  split; rintro ⟨c, hc⟩; refine ⟨-c, λ ω hω, _⟩,
+  { rw mem_upper_bounds at hc,
+    refine neg_le.2 (hc _ _),
+    simpa only [pi.neg_apply, set.mem_range, neg_inj] },
+  { rw mem_lower_bounds at hc,
+    simp_rw [set.mem_range, pi.neg_apply, neg_eq_iff_neg_eq, eq_comm] at hω,
+    refine le_neg.1 (hc _ _),
+    simpa only [set.mem_range] }
+end
+
+lemma martingale.ae_not_tendsto_at_top_at_top [is_finite_measure μ]
+  (hf : martingale f ℱ μ) (hbdd : ∀ᵐ ω ∂μ, ∀ i, |f (i + 1) ω - f i ω| ≤ R) :
+  ∀ᵐ ω ∂μ, ¬ tendsto (λ n, f n ω) at_top at_top :=
+by filter_upwards [hf.bdd_above_range_iff_bdd_below_range hbdd] with ω hω htop using
+    unbounded_of_tendsto_at_top htop (hω.2 $ bdd_below_range_of_tendsto_at_top_at_top htop)
+
+lemma martingale.ae_not_tendsto_at_top_at_bot [is_finite_measure μ]
+  (hf : martingale f ℱ μ) (hbdd : ∀ᵐ ω ∂μ, ∀ i, |f (i + 1) ω - f i ω| ≤ R) :
+  ∀ᵐ ω ∂μ, ¬ tendsto (λ n, f n ω) at_top at_bot :=
+by filter_upwards [hf.bdd_above_range_iff_bdd_below_range hbdd] with ω hω htop using
+    unbounded_of_tendsto_at_bot htop (hω.1 $ bdd_above_range_of_tendsto_at_top_at_bot htop)
+
+namespace borel_cantelli
+
+/-- Auxiliary definition required to prove Lévy's generalization of the Borel-Cantelli lemmas for
+which we will take the martingale part. -/
+noncomputable
+def process (s : ℕ → set Ω) (n : ℕ) : Ω → ℝ :=
+∑ k in finset.range n, (s (k + 1)).indicator 1
+
+variables {s : ℕ → set Ω}
+
+lemma process_zero : process s 0 = 0 :=
+by rw [process, finset.range_zero, finset.sum_empty]
+
+lemma adapted_process (hs : ∀ n, measurable_set[ℱ n] (s n)) :
+  adapted ℱ (process s) :=
+λ n, finset.strongly_measurable_sum' _ $ λ k hk, strongly_measurable_one.indicator $
+  ℱ.mono (finset.mem_range.1 hk) _ $ hs _
+
+lemma martingale_part_process_ae_eq (ℱ : filtration ℕ m0) (s : ℕ → set Ω) (n : ℕ) :
+  martingale_part ℱ μ (process s) n =ᵐ[μ]
+  ∑ k in finset.range n, ((s (k + 1)).indicator 1 - μ[(s (k + 1)).indicator 1 | ℱ k]) :=
+begin
+  simp only [martingale_part_eq_sum, process_zero, zero_add],
+  refine eventually_eq_sum (λ k hk, _),
+  simp only [process, finset.sum_range_succ_sub_sum],
+end
+
+lemma martingale_part_process_ae_eq' (ℱ : filtration ℕ m0) (s : ℕ → set Ω) :
+  ∀ᵐ ω ∂μ, ∀ n, martingale_part ℱ μ (process s) n ω =
+  ∑ k in finset.range n, ((s (k + 1)).indicator 1 - μ[(s (k + 1)).indicator 1 | ℱ k]) ω :=
+begin
+  rw ae_all_iff,
+  intro n,
+  convert martingale_part_process_ae_eq ℱ s n,
+  ext ω,
+  rw finset.sum_apply,
+end
+
+variables [is_finite_measure μ]
+
+lemma integrable_process (hs : ∀ n, measurable_set[ℱ n] (s n)) (n : ℕ) :
+  integrable (process s n) μ :=
+integrable_finset_sum' _ $ λ k hk, integrable_on.indicator (integrable_const 1) $ ℱ.le _ _ $ hs _
+
+lemma martingale_part_process_diff_le (hs : ∀ n, measurable_set[ℱ n] (s n)) (n : ℕ) :
+  ∀ᵐ ω ∂μ, |martingale_part ℱ μ (process s) (n + 1) ω - martingale_part ℱ μ (process s) n ω| ≤ 1 :=
```

Comment:
Could this be a consequence of a more general result of the form "if a process has bounded increments, then its martingale part also has bounded increments"?

Reply:
I've added `martingale_part_bdd_difference` now